opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,676 @@
+  result->total_elapsed_time = this->now() - action_start_time_;
+  result->num_retries = num_retries_;
+  publishZeroVelocity();
+  following_action_server_->terminate_current(result);
+  dynamic_pose_sub_.reset();
+}
+
+bool FollowingServer::approachObject(
+  geometry_msgs::msg::PoseStamped & object_pose, const std::string & target_frame)
+{
+  rclcpp::Rate loop_rate(controller_frequency_);
+  while (rclcpp::ok()) {
+    // Update the iteration start time, used for get robot position, transformation and control
+    iteration_start_time_ = this->now();
+
+    publishFollowingFeedback(FollowObject::Feedback::CONTROLLING);
+
+    // Stop if cancelled/preempted
+    if (checkAndWarnIfCancelled<FollowObject>(following_action_server_, "follow_object") ||
+      checkAndWarnIfPreempted<FollowObject>(following_action_server_, "follow_object"))
+    {
+      return false;
+    }
+
+    // If we have a target frame_id from the goal, use that instead of pose detection
+    if (!target_frame.empty()) {
+      if (!getFramePose(target_frame, object_pose)) {
+        throw opennav_docking_core::FailedToDetectDock(
+          "Failed to get pose in target frame: " + target_frame);
+      }
+    } else {
+      // Otherwise, use the traditional pose detection from topic
+      if (!getRefinedPose(object_pose)) {
+        throw opennav_docking_core::FailedToDetectDock("Failed object detection");
+      }
+    }
+
+    // Get the pose at the distance we want to maintain from the object
+    // Stop and report success if goal is reached
+    auto target_pose = getPoseAtDistance(object_pose, desired_distance_);
+    if (isGoalReached(target_pose)) {
+      return true;
+    }
+
+    // The control law can get jittery when close to the end when atan2's can explode.
+    // Thus, we reduce the desired distance by a small amount so that the robot never
+    // gets to the end of the spiral before its at the desired distance to stop the
+    // following procedure.
+    const double backward_projection = 0.25;
+    const double effective_distance = desired_distance_ - backward_projection;
+    target_pose = getPoseAtDistance(object_pose, effective_distance);
+
+    // ... and transform the target_pose into base_frame
+    try {
+      tf2_buffer_->transform(
+        target_pose, target_pose, base_frame_, tf2::durationFromSec(transform_tolerance_));
+    } catch (const tf2::TransformException & ex) {
+      RCLCPP_WARN(get_logger(), "Failed to transform target pose: %s", ex.what());
+      return false;
+    }
+
+    // If the object is behind the robot, we reverse the control
+    auto robot_pose = getRobotPoseInFrame(target_pose.header.frame_id);
+    const double dist = std::hypot(
+      robot_pose.pose.position.x - target_pose.pose.position.x,
+      robot_pose.pose.position.y - target_pose.pose.position.y);
+    bool reversing = allow_backward_ &&
+      dist < desired_distance_ && target_pose.pose.position.x < robot_pose.pose.position.x;
```

Comment:
I don't quite understand this point `dist < desired_distance_`. Why if we're closer only to the desired distance do we allow reversing? 

I might also be misunderstanding the reversing topic. Is this a temporary reversing or is this now just executing the task in the reverse direction?